File: examples/notif-wake/go/notif-wake.go
// Copyright 2021, 2022 Nokia
// Licensed under the BSD 3-Clause Clear License.
// SPDX-License-Identifier: BSD-3-Clause-Clear

// Run:
// go run ./...

package main

import (
	"context"
	"flag"
	"fmt"
	"net/http"
	"net/url"
	"os"
	"sync"
	"time"

	"github.com/nokia/restful"
)

var (
	msisdn       = flag.String("msisdn", "", "MSISDN, i.e. phone number of the device")
	nef          = flag.String("nef", "", "URL of NEF")
	tokenURL     = flag.String("tokenurl", "", "URL of authorization server serving access token")
	clientID     = flag.String("clientid", "", "Client ID")
	clientSecret = flag.String("clientsecret", "", "Client Secret")
	af           = flag.String("af", "", "Application Function ID")
	notifyURL    = flag.String("notifyurl", "", "URL where you expect the notifications")
)

type MonitoringEventSubscription struct { // Simplified
	Self                    string `json:"self,omitempty"`
	Msisdn                  string `json:"msisdn,omitempty"`
	NotificationDestination string `json:"notificationDestination"`
	MonitoringType          string `json:"monitoringType"`
	MaximumNumberOfReports  uint32 `json:"maximumNumberOfReports,omitempty"`
	ReachabilityType        string `json:"reachabilityType,omitempty"`
}

type MonitoringNotification struct {
	Subscription string `json:"subscription"`
}

func checkFlags() {
	flag.Parse()
	flag.VisitAll(func(f *flag.Flag) {
		if f.Value.String() == "" {
			println("Argument not set: -" + f.Name)
			flag.Usage()
			os.Exit(1)
		}
	})
}

func panicIf(err error) {
	if err != nil {
		panic(err)
	}
}

// Server

type server struct {
	client *restful.Client
	wg     *sync.WaitGroup
}

func (s *server) notify(ctx context.Context, data *MonitoringNotification) error {
	if data.Subscription == "" {
		return restful.NewError(nil, http.StatusBadRequest, "missing subscription")
	}
	defer s.wg.Done()
	fmt.Println("Notification:", data.Subscription)
	panicIf(s.client.Delete(ctx, data.Subscription))
	return nil
}

func startServer(client *restful.Client, wg *sync.WaitGroup) {
	s := server{client: client, wg: wg}
	s.wg.Add(1)
	r := restful.NewRouter()
	r.HandleFunc("/", s.notify)
	go r.ListenAndServe(":8080")
}

// Client

func createClient(root string) *restful.Client {
	return restful.NewClient().
		Root(root).
		Retry(3, 500*time.Millisecond, 2*time.Second).
		Timeout(10 * time.Second)
}

// Get access token

type tokenResp struct {
	AccessToken string `json:"access_token"`
}

func getAccessToken() (string, error) {
	c := createClient("")

	req := url.Values{}
	req.Set("grant_type", "client_credentials")
	req.Set("client_id", *clientID)
	req.Set("client_secret", *clientSecret)
	var resp tokenResp

	_, err := c.PostForm(context.Background(), *tokenURL, req, &resp)
	return resp.AccessToken, err
}

func subscribe(client *restful.Client, token string) string {
	ctx := context.Background()

	data := MonitoringEventSubscription{Msisdn: *msisdn, NotificationDestination: *notifyURL, MonitoringType: "UE_REACHABILITY", MaximumNumberOfReports: 1, ReachabilityType: "DATA"}
	var created MonitoringEventSubscription
	headers := http.Header{}
	headers.Set("Authorization", "Bearer "+token)
	resp, err := client.SendRecv2xx(ctx, http.MethodPost, "/3gpp-monitoring-event/v1/"+*af+"/subscriptions", headers, &data, &created)
	panicIf(err)
	resource, err := resp.Location()
	fmt.Printf("Created: %+v\n", created)
	fmt.Printf("Resource URL: %v\n", resource)
	return resource.String()
}

func query(client *restful.Client, resource, token string) {
	ctx := context.Background()

	var created MonitoringEventSubscription
	headers := http.Header{}
	headers.Set("Authorization", "Bearer "+token)
	_, err := client.SendRecv2xx(ctx, http.MethodGet, resource, headers, nil, &created)
	panicIf(err)
	fmt.Printf("Queried: %+v\n", created)
}

// Main

func main() {
	checkFlags()

	token, err := getAccessToken()
	panicIf(err)

	var wg sync.WaitGroup
	client := createClient(*nef)
	startServer(client, &wg)
	resource := subscribe(client, token)
	query(client, resource, token)
	wg.Wait()
}

File: examples/notif-wake/python/notif-wake.py
#!/usr/bin/env python3

# Copyright 2021, 2022 Nokia
# Licensed under the BSD 3-Clause Clear License.
# SPDX-License-Identifier: BSD-3-Clause-Clear

# Run:
# pip install -r requirements.txt
# ./notif-wake.py

import argparse
import requests
import threading
import urllib

from requests.sessions import Session
from fastapi import FastAPI
from pydantic import BaseModel
import uvicorn

def parse_args():
	arg_parser = argparse.ArgumentParser()
	arg_parser.add_argument('-m', '--msisdn', action='store', type=str, required=True, help="MSISDN, i.e. phone number of the device")
	arg_parser.add_argument('-n', '--nef', action='store', type=str, required=True, help="URL of NEF")
	arg_parser.add_argument('-t', '--tokenurl', action='store', type=str, required=True, help="URL of authorization server serving access token")
	arg_parser.add_argument('-c', '--clientid', action='store', type=str, required=True, help="Client ID")
	arg_parser.add_argument('-s', '--clientsecret', action='store', type=str, required=True, help="Client secret")
	arg_parser.add_argument('-a', '--af', action='store', type=str, required=True, help="Application Function ID")
	arg_parser.add_argument('-N', '--notifyurl', action='store', type=str, required=True, help="URL where you expect the notifications")
	return arg_parser.parse_args()

class Client:
	def __init__(self, args: argparse.Namespace):
		self.args = args
		self.client = requests.Session()
		self.client.headers["Authorization"] = "Bearer " + self._get_token()

	def _get_token(self) -> str:
		data={"grant_type": "client_credentials", "client_id": self.args.clientid, "client_secret": self.args.clientsecret}
		resp = requests.post(url=self.args.tokenurl, data=data)
		access_token = resp.json()["access_token"]
		return access_token

	def subscribe(self)->str:
		data = {"msisdn": self.args.msisdn, "notificationDestination": self.args.notifyurl, "monitoringType": "UE_REACHABILITY", "reachabilityType": "DATA", "maximumNumberOfReports": 1}
		response = self.client.post(self.args.nef + '/3gpp-monitoring-event/v1/'+self.args.af+'/subscriptions', json=data)
		response.raise_for_status()
		resource = response.headers['Location']
		print("Created:", response.json())
		print("Resource URL:", resource)
		return resource

	def query(self, resource: str):
		response = self.client.get(resource)
		print("Queried:", response.json())

	def delete(self, resource: str):
		self.client.delete(resource)


class MonitoringNotification(BaseModel):
	subscription: str


args = parse_args()
client = Client(args)

app = FastAPI()

def start_server()->threading.Thread:
	server = threading.Thread(target=uvicorn.run, args=(app,))
	server.start()
	return server

@app.post("/")
def notification(notif: MonitoringNotification):
	print("Notification for:", notif.subscription)
	client.delete(notif.subscription)

if __name__ == "__main__":
	start_server()
	resource = client.subscribe()
	client.query(resource)
